tools/release/test_lp_build_monitor_recipe.py

```diff
@@ -0,0 +1,262 @@
+            lp_build_monitor_recipe.monitor_retry_builds(
+                MagicMock(), MagicMock()
+            )
+
+        # we updated till the build reported a success
+        self.assertEqual(build_mock.lp_refresh.call_count, 1)
+        # each time a failure was detected, the build was retried
+        self.assertEqual(build_mock.retry.call_count, 0)
+        # we don't update a build that can't be retried at all
+        self.assertEqual(len(build_status_evolution), 1)
+
+
+class TestMain(unittest.TestCase):
+    @patch("time.sleep", new=MagicMock())
+    @patch(
+        "lp_build_monitor_recipe.get_date_utc_now",
+        new=MagicMock(return_value=10),
+    )
+    @patch("lp_build_monitor_recipe.print", new=MagicMock())
+    @patch("lp_build_monitor_recipe.get_build_recipe")
+    def test_success(self, get_build_recipe_mock):
+        recipe_mock = get_build_recipe_mock()
+        # we are asking this to build on 3 platforms
+        recipe_mock.distroseries = ["Xenial", "Bionic", "Jammy"]
+        # every build starts and pends for some time
+        pending_build_info = [
+            [
+                {"distroseries": "Xenial"},
+                {"distroseries": "Bionic"},
+                {"distroseries": "Jammy"},
+            ],
+            [
+                {"distroseries": "Xenial"},
+            ],
+            [],
+        ]
+        recipe_mock.getPendingBuildInfo.side_effect = pending_build_info
+        # the builds started and some failed and must be retried
+        build_mock = MagicMock()
+        build_mock.date_first_dispatched = 11  # more than now, 10
+        build_mock.can_be_retried = True
+        # A build is updated via the lp_refresh function, lets do the same
+        # here but inject our test values
+        build_status_evolution = [
+            "Successfully built",
+            "Successfully built",
+            "Successfully built",
+            "Uploading build",
+            "Gathering build output",
+            "Currently building",
+            "Failed to build",
+        ]
+
+        def lp_refresh_side_effect():
+            if build_status_evolution:
+                build_mock.buildstate = build_status_evolution.pop()
+
+        build_mock.lp_refresh.side_effect = lp_refresh_side_effect
+        recipe_mock.builds = [build_mock, build_mock, build_mock]
+
+        lp_build_monitor_recipe.main(["checkbox", "--recipe", "some_recipe"])
+
+        self.assertEqual(recipe_mock.requestBuild.call_count, 3)
+        # after 3 calls all builds were started correctly
+        self.assertEqual(recipe_mock.getPendingBuildInfo.call_count, 3)
+        # only 1 build failed and was retried
+        self.assertEqual(build_mock.retry.call_count, 1)
+
+    @patch("time.sleep", new=MagicMock())
+    @patch(
+        "lp_build_monitor_recipe.get_date_utc_now",
+        new=MagicMock(return_value=10),
+    )
+    @patch("lp_build_monitor_recipe.print", new=MagicMock())
```

Comment:
there's no check to what the print was called with, so maybe the mock is not needed at all?